src/dra/5D874.c

```diff
@@ -2143,9 +2143,171 @@ void func_80123A60(Entity* entity) {
 
 INCLUDE_ASM("asm/dra/nonmatchings/5D874", func_80123B40);
 
-INCLUDE_ASM("asm/dra/nonmatchings/5D874", func_80123F78);
+void func_80123F78(Entity* entity) {
+    if (D_800973FC == 0) {
+        D_80097F3F = 0;
+        func_80106590(entity);
+        return;
+    }
 
-INCLUDE_ASM("asm/dra/nonmatchings/5D874", func_80124164);
+    switch (entity->step) {
+    case 0:
+        entity->unk34 = 0x0C030000;
+        if (PLAYER.animationSet != 1) {
+            func_80106590(entity);
+            break;
+        }
+        entity->animationSet = 1;
+        entity->animationFrame = PLAYER.animationFrame;
+        entity->unk5A = 0xD;
+        entity->unk19 = PLAYER.unk19;
+        entity->blendMode = 0x30;
+        entity->palette = 0x815F;
+        entity->zPriority = PLAYER.zPriority - 2;
+        entity->facing = PLAYER.facing;
+        do { // TODO: !FAKE
+        } while (0);
+        D_80097F4A = 0x80;
+        D_80097F46 = 0x80;
+        D_80097F42 = 0x80;
+        D_80097F3E = 0x80;
+        D_80097F49 = 0x80;
+        D_80097F45 = 0x80;
+        D_80097F41 = 0x80;
+        D_80097F3D = 0x80;
+        D_80097F48 = 0x80;
+        D_80097F44 = 0x80;
+        D_80097F40 = 0x80;
+        D_80097F3C = 0x80;
+        D_80097F3F = 1;
+        do { // TODO: !FAKE
+        } while (0);
+        entity->step++;
+        break;
+
+    case 1:
+        D_80097F4A += 255;
+        if (D_80097F4A < 4) {
+            D_80097F3F = 0;
+            func_80106590(entity);
+            break;
+        }
+        D_80097F46 = D_80097F4A;
+        D_80097F42 = D_80097F4A;
+        D_80097F3E = D_80097F4A;
+        D_80097F49 = D_80097F4A;
+        D_80097F45 = D_80097F4A;
+        D_80097F41 = D_80097F4A;
+        D_80097F3D = D_80097F4A;
+        D_80097F48 = D_80097F4A;
+        D_80097F44 = D_80097F4A;
+        D_80097F40 = D_80097F4A;
+        D_80097F3C = D_80097F4A;
+        break;
+    }
+}
+
+void func_80124164(POLY_GT4* poly, s32 arg1, s32 arg2, s32 arg3, s32 arg4) {
+    s16 temp = arg2 - arg3;
+    s16 temp2 = arg2 + arg3;
+    s32 temp3;
+
+    poly->y1 = temp;
+    poly->y0 = temp;
+
+    poly->y3 = temp2;
+    poly->y2 = temp2;
+
+    if (poly->y0 < 0) {
+        poly->y1 = 0;
+        poly->y0 = 0;
+    }
+
+    if (poly->y0 > 240) {
+        poly->y3 = 240;
+        poly->y2 = 240;
+    }
+
+    if (arg4 == 0) {
+        temp3 = arg1 * D_800AE270[0];
+        if (temp3 < 0) {
+            temp3 += 255;
+        }
+        poly->r3 = (u32)temp3 >> 8;
```

Comment:
You can remove those `(u32)` casts by changing the type of `temp3` into `u32`. Also I think `temp3` can be renamed as `color`, `colorUnit` or `colorChannel`